### PR TITLE
Model Fitting

### LINKED ISSUE #414: Re-implement spectral fitting
- Create new `ListModel` to contain parameter items.
- (Astropy) implement scipy's curve fit as an astropy fitter to provide parameter errors
- Astropy compound model arithmetic

### PR DESCRIPTION
**Pull request by robelgeda**
Closes #414 
This PR enables model fitting via SpecUtil fit_lines function. **For testing**, Please make sure to update your SpecUtils.
screen shot 2018-11-08 at 4 20 53 am

### CODE REVIEW

**Comment by drdavella** [comment]
This looks good. Just a few quick comments after an initial look:

* When I click "New Model" the new "Fittable Model Spectrum" should automatically appear in the sidebar, and should also automatically be selected/displayed in the plot
* It's not at all clear to me right now that it's necessary to manually set model parameters before fitting. @brechmos-stsci tells me that automatic parameter estimation will be added to `specutils` shortly, though, so I guess we don't need to do anything right now.
* There is currently no indication of goodness-of-fit or whether the fitting even happened. Can we add some kind of indicator for this?

**Comment by robelgeda** [comment]
This looks good. Just a few quick comments after an initial look:

> * When I click "New Model" the new "Fittable Model Spectrum" should automatically appear in the sidebar, and should also automatically be selected/displayed in the plot
Agreed, I will try to add it this round, This thing is going to be significantly different by the next release. 

> * It's not at all clear to me right now that it's necessary to manually set model parameters before fitting. @brechmos-stsci tells me that automatic parameter estimation will be added to `specutils` shortly, though, so I guess we don't need to do anything right now.
This is correct, some work is going into the SpecUtils side.

> * There is currently no indication of goodness-of-fit or whether the fitting even happened. Can we add some kind of indicator for this?
I am not sure how to do this. I am open to recomendations

**Comment by javerbukh** [comment]
Experienced a crash doing the following:
1. Load specviz with COS_FUV.fits
2. Model Editor -> Create New Model
3. \+ -> Gaussian1D
4. Fit Model -> OK
5. Error. Here is the traceback
```Traceback (most recent call last):
  File "/Users/javerbukh/Documents/specviz/specviz/plugins/model_editor/model_editor.py", line 310, in _on_fit_clicked
    epsilon=epsilon)
TypeError: fit_lines() got an unexpected keyword argument 'maxiter'
Abort trap: 6```

**Comment by drdavella** [comment]
@robelgeda @brechmos-stsci is this failure because this feature relies on an unmerged PR to `specutils`?

**Comment by brechmos-stsci** [comment]
@drdavella @javerbukh Yes, currently you must use my branch from specutils.  Hopefully the specutils branch will be merged today and then you will be able to use master.

**Comment by nluetzge** [comment]
Got a crash after I fitted the model and then tried to change the units from Angstrom to Herz...Here is the error: I guess this is hard for the model because it actually changes the shape...
```
Traceback (most recent call last):
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/unit_change/unit_change_dialog.py", line 268, in on_accepted
    if not plot_data_item.is_data_unit_compatible(data_unit_formatted):
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/core/items.py", line 143, in is_data_unit_compatible
    return (self.data_item.flux.unit == "" or
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/model_editor/items.py", line 30, in flux
    model_flux = result(self.model_spectral_axis.value) * model_flux_units
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/model_editor/items.py", line 43, in model_spectral_axis
    return self.spectral_axis.to(model_spectral_units)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/quantity.py", line 847, in to
    return self._new_view(self._to_value(unit, equivalencies), unit)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/quantity.py", line 819, in _to_value
    equivalencies=equivalencies)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/core.py", line 965, in to
    return self._get_converter(other, equivalencies=equivalencies)(value)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/core.py", line 899, in _get_converter
    raise exc
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/core.py", line 885, in _get_converter
    self, other, self._normalize_equivalencies(equivalencies))
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/astropy/units/core.py", line 869, in _apply_equivalencies
    unit_str, other_str))
astropy.units.core.UnitConversionError: 'Angstrom' (length) and 'Hz' (frequency) are not convertible
Abort trap: 6
```

**Comment by robelgeda** [comment]
Actually that was in response to the last one. This one is different, which plot was selected when you got this error?

**Comment by robelgeda** [comment]
Thanks @nluetzge !  That was an incompatibility problem with the arithmetics editor. The bug you posted however is a SpecViz problem. If you divide two spectra, the resulting spectra will be unit-less. SpecViz does not know how to plot data without units.

File: specviz/plugins/model_editor/model_editor.py [hunk 1]
```diff
@@ -101,19 +145,60 @@ def _on_remove_model(self):
...
 
-    def _add_fittable_model(self, model):
-        idx = self.model_tree_view.model().add_model(model())
+    def _add_fittable_model(self, model_type):
+        if issubclass(model_type, MODELS['Polynomial1D']):
+            text, ok = QInputDialog.getInt(self, 'Polynomial1D',
+                                           'Enter Polynomial1D degree:')
+            if ok:
+                # Create a new astropy model class definition with the degree
+                # data already embedded
```
Comment (drdavella) [inline specviz/plugins/model_editor/model_editor.py]: Why is it necessary to do this? Why can't we just create an instance of Astropy's `Polynomial1D` here?

**Pull request merged**